packages/cli/src/services/redis/RedisServiceStreamConsumer.ts

```diff
@@ -0,0 +1,92 @@
+import { Service } from 'typedi';
+import { LoggerProxy } from 'n8n-workflow';
+import { RedisServiceBaseReceiver } from './RedisServiceBaseClasses';
+
+type LastId = string;
+
+type StreamName = string;
+
+type StreamDetails = {
+	lastId: LastId;
+	pollingInterval: number;
+	waiter: NodeJS.Timeout | undefined;
+};
+
+@Service()
+export class RedisServiceStreamConsumer extends RedisServiceBaseReceiver {
+	// while actively listening, the stream name and last id are stored here
+	// removing the entry will stop the listener
+	streams: Map<StreamName, StreamDetails> = new Map();
+
+	async init(): Promise<void> {
+		await super.init('consumer');
+	}
+
+	async listenToStream(stream: StreamName, lastId = '$'): Promise<void> {
+		if (!this.redisClient) {
+			await this.init();
+		}
+		LoggerProxy.debug(`Redis client now listening to stream ${stream} starting with id ${lastId}`);
+		this.setLastId(stream, lastId);
+		let failedRuns = 0;
+		while (this.streams.has(stream)) {
+			const results = await this.redisClient?.xread('BLOCK', 0, 'STREAMS', stream, lastId);
+			if (results) {
+				const [_key, messages] = results[0];
+				if (messages.length > 0) {
+					messages.forEach(([id, message]) => {
+						this.messageHandlers.forEach((handler) => handler(stream, id, message));
+					});
+					// Pass the last id of the results to the next round.
+					lastId = messages[messages.length - 1][0];
+					this.setLastId(stream, lastId);
+				}
+			} else {
+				failedRuns++;
+			}
+			if (failedRuns > 10) {
+				LoggerProxy.warn(
+					`Redis stream ${stream} is not available (failed ${failedRuns} times). Stopping listening.`,
```

Comment:
Maybe we should emit an event here, so that n8n can exit as it's probably out of sync, or at least try reloading from DB WDYT?

Reply:
Yes we need to react in some way here. I've not quite made up my mind yet in what situations this error could occur and how to handle them. It's just a safeguard to at least exit the while loop for now. I think a rewrite into a setTimeout() function may be in order here anyway....